File: src/NubeSync.Client/Data/NubeOperation.cs
using System;

namespace NubeSync.Client.Data
{
    public class NubeOperation
    {
        public NubeOperation()
        {
            Id = Guid.NewGuid().ToString();
            CreatedAt = DateTimeOffset.Now;
        }

        public DateTimeOffset CreatedAt { get; set; }

        public string Id { get; set; }

        public string ItemId { get; set; } = string.Empty;

        public string? OldValue { get; set; }

        public string? Property { get; set; }

        public string TableName { get; set; } = string.Empty;

        public OperationType Type { get; set; }

        public string? Value { get; set; }

        public override string ToString()
        {
            return $"Id {Id}, {Type} {Property} in table {TableName} for item {ItemId} with value {Value} (old: {OldValue})";
        }
    }
}

File: samples/Clients/Flutter/nubesyncflutter/lib/_nubeSync/nube_operation.dart
class NubeOperation {
  String? id;
  DateTime? createdAt;
  String itemId;
  String? oldValue;
  String property;
  String tableName;
  OperationType type;
  String? value;

  NubeOperation(
      {this.createdAt,
      required this.itemId,
      required this.value,
      required this.property,
      required this.tableName,
      required this.type,
      this.oldValue,
      }) {
        createdAt = DateTime.now();
      }

  Map<String, dynamic> toMap() {
    var map = <String, dynamic>{
      'createdAt': createdAt != null
          ? createdAt!.toIso8601String()
          : DateTime.now().toString(),
      'itemId': itemId,
      'oldValue': oldValue,
      'property': property,
      'tableName': tableName,
      'type': type.index,
      'value': value
    };

    if (id != null) {
      map['id'] = id;
    }

    return map;
  }

  factory NubeOperation.fromMap(Map<String, dynamic> map) {
    var operation = NubeOperation(
      createdAt: DateTime.parse(map['createdAt']), 
      itemId: map['itemId'], 
      oldValue: map['oldValue'], 
      property: map['property'], 
      tableName: map['tableName'],
      type: OperationType.values[map['type']], 
      value: map['value'], 
    );

    operation.id = map['id'];
    return operation;
  }
}

enum OperationType { added, modified, deleted }